Math/NumberTheory/QuadraticIntegers.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE DataKinds      #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE PolyKinds #-}
+
+module Math.NumberTheory.QuadraticIntegers (
+  QuadraticInteger(..),
+  Sign(..),
+  discriminant,
+  conjugate,
+  norm,
+  trace,
+  quotRemQ,
+  quotQ,
+  remQ,
+  divModQ,
+  divQ,
+  modQ,
+) where
+
+import GHC.TypeNats.Compat
+import Data.Proxy
+
+infix 5 :+
+
+--------------------------------------------------------------
+-- Type level representation of sign
+  
+data Sign = Pos | Neg
+
+class KnownSign (x :: Sign) where
+  signVal :: proxy x -> Sign
+
+instance KnownSign 'Pos where
+  signVal _ = Pos
+
+instance KnownSign 'Neg where
+  signVal _ = Neg
+
+-- | Takes a quadratic integer and returns its sign as an integer
+signToInt :: forall s d . KnownSign s => QuadraticInteger s d -> Integer
+signToInt _ =
+  case signVal (Proxy :: Proxy s) of
+    Pos -> 1
+    Neg -> -1
+  
+---------------------------------------------------------------
+-- Data definitions for quadratic integers
+
+data QuadraticInteger (sign :: Sign) (d :: Nat) =
+  (:+) { rationalComp :: !Integer, irrationalComp :: !Integer } deriving (Eq)
+
+
+{- If the discriminant of a field is equivalent to 1 mod 4
+   then the basis for the ring of integers is strictly
+   larger than a + bx. Need to represent this other basis -}
+
+data BasisType = Integral | Fractional
+
+
+-- | Takes a quadratic int and returns its basis type
+basisType :: forall s d . (KnownNat d, KnownSign s)
+          => QuadraticInteger s d
+          -> BasisType
+basisType z
+  | disc `mod` 4 == 1 = Fractional
+  | otherwise         = Integral
+  where disc = discriminant z
+
+
+instance Show (QuadraticInteger s d) where
+    show (a :+ b)
+        | b == 0     = show a
+        | a == 0     = s ++ b'
+        | otherwise  = show a ++ op ++ b'
+        where
+            b' = if abs b == 1 then "x" else show (abs b) ++ "*x"
+            op = if b > 0 then "+" else "-"
+            s  = if b > 0 then "" else "-"
+
+instance forall s d . (KnownNat d, KnownSign s) => Num (QuadraticInteger s d) where
+  (+) (a :+ b) (c :+ d) = (a+c) :+ (b+d)
+
+  (*) x@(a :+ b) (c :+ d) =
+    case basisType x of
+      Integral -> (a*c) + (b*d)*(discriminant x) :+ (a*d + c*b)
+      Fractional -> (a*c + b*d*((discriminant x)-1) `div` 4) :+ (a*d + b*c + b*d)
+
+  abs = undefined
```

Comment:
It is better to give a trivial definition than leave `abs` and `signum` undefined. Type class `Num` requires that `x == abs x * signum x`, so one can put `abs x = x` and `signum x = 1 :+ 0`, for instance. 

Another (and better) option is to define `abs` as a map to factor ring by ring units (6 units for Eisenstein numbers, 4 units for Gaussian ones and 2 in other cases). Cf. https://github.com/cartazio/arithmoi/blob/master/Math/NumberTheory/GaussianIntegers.hs#L68